### DISCUSSION TITLE
[Support] precognition requests via middleware

### DISCUSSION THREAD

**Discussion by zdravstvuy1714** [spatie/laravel-data#343] [Ideas]
implement support for https://github.com/laravel/framework/pull/44339 Laravel feature

**Comment by ragulka** [comment]
The following workaround adds Precognition support to Laravel Data. It intercepts the pipeline if the payload is a precognitive request, then filters the rules based on the precognition headers and validates the payload. Otherwise, it will continue as usual.

`PrecognitivelyValidatePropertiesDataPipe.php`:

```php
use Illuminate\Http\Request;
use Illuminate\Support\Collection;
use Illuminate\Support\Facades\Validator;
use Spatie\LaravelData\DataPipes\ValidatePropertiesDataPipe;
use Spatie\LaravelData\Support\DataClass;

class PrecognitivelyValidatePropertiesDataPipe extends ValidatePropertiesDataPipe
{
    public function handle(mixed $payload, DataClass $class, Collection $properties): Collection
    {
        if ($payload instanceof Request && $payload->isPrecognitive()) {
            Validator::make(
                $properties->toArray(),
                $payload->filterPrecognitiveRules(($class->name)::getValidationRules($properties->toArray())),
            )->validate();

            return $properties;
        }

        return parent::handle($payload, $class, $properties);
    }
}
```

`AppServiceProvider.php`:

```php
class AppServiceProvider extends ServiceProvider
{
    public function boot(): void
    {
        $this->app->bind(ValidatePropertiesDataPipe::class, PrecognitivelyValidatePropertiesDataPipe::class);
    }
}
```

**Reply by rubenvanassche** [reply]
Is this something we want to add to the package?

**Reply by ragulka** [reply]
Note that what I have above is not enough... it does not stop the execution after validation. The Data object will still be created, but it will only receive the fields that are present in the `Precognition-Validate-Only` header. If the data object has a required field that is not present in the header, the creation will fail and the endpoint will result in a 500 server error.

We need some way to stop the pipeline execution after validation during precognitive requests, but I haven't figured it out, yet.

**Reply by ragulka** [reply]
An alternative approach is to validate in the `LaravelDataServicerProvider`, when binding the data class:

```php
$this->app->beforeResolving(BaseData::class, function ($class, $parameters, $app) {
    if ($app->has($class)) {
        return;
    }

    $app->bind(
        $class,
        function ($container)  use ($class) {
            $request = $container['request'];

            if ($request->isPrecognitive()) {
                return Validator::make(
                    $request->all(),
                    $request->filterPrecognitiveRules(($class)::getValidationRules($request->all())),
                )->validate();
            }

            return $class::from($container['request']);
        }
    );
});
```

However, this approach will skip the pipes that normally occur _before_ request validation, as well as the `prepareForPipeline` method, which is why I'm not a huge fan of this. But it does effectively stop execution after validation.

**Reply by rubenvanassche** [reply]
Hmm I think breaking out of the pipeline should be the way to go. The alternative approach feels a bit odd, I'll add it on the list for v4.

**Reply by ragulka** [reply]
I agree. Though, the only way I could break out the pipeline in the current version was to do one of the following in [`DataFromSomethingResolver::execute()`](https://github.com/spatie/laravel-data/blob/5e170162c573085b7529cd1cd8881201f22b5110/src/Resolvers/DataFromSomethingResolver.php#L46-L48): 

1. `return` early with a "blank" data object, ie `MyData::from(MyData::blank())`. This is flaky, as it's not guaranteed that `blank` returns valid props for a new data object - it depends on the data class implementation and there may be required fields, or compound types, so this was not a very reliable way to do it.

2. `throw` a specific exception class, ie `DataValidatedPrecognitivelyException()`, then "render" it in `ExceptionHandler` as a successful response:
```php
$this->renderable(function (DataValidatedPrecognitivelyException $e, Request $request) {
    return response('ok', 204)->header('Precognition', 'true');
});
```

Needless to say, both approaches are quite hacky, although the latter is at least reliable.

**Reply by jameshulse** [reply]
Has there been progress on this? Thanks.

**Reply by kirkita** [reply]
Yah, I just encountered this one. 🙏

**Reply by rubenvanassche** [reply]
Not working on it right now, this is something for v5 since we're going to take a look at validation. Atm I'm working on typescript-transformer v3 and fixing data v4 bugs so this probably will take a few months before I even can start working on v5.

If someone finds a non-hacky, non-breaking change solution to the problem, I'll gladly merge it into v4. Otherwise it will have to wait until v5.

**Reply by jameshulse** [reply]
@ragulka is 99% of the way there with his update to the `ValidatePropertiesDataPipe`. The way Precognition exits early in Laravel seems to be via an `->after(...)` Validation hook. They provide a default implementation which calls `abort(204, ...)` if the validation succeeds.

So updating the pipeline with the same logic, it now looks something like:

```diff
class PrecognitivelyValidatePropertiesDataPipe extends ValidatePropertiesDataPipe
{
    /**
     * @throws ValidationException
     */
    public function handle(mixed $payload, DataClass $class, Collection $properties): Collection
    {
        if ($payload instanceof Request && $payload->isPrecognitive()) {
            Validator::make(
                $properties->toArray(),
                $payload->filterPrecognitiveRules(($class->name)::getValidationRules($properties->toArray())),
            )
+               ->after(Precognition::afterValidationHook($payload))
                ->validate();
        }

        return parent::handle($payload, $class, $properties);
    }
}

```

In my basic testing - this means I get an HTTP 204 when validation passes, and a 429 when it fails. This seems to be the same behaviour with standard Laravel.

So I think we are really close to having a snippet that could drop into the package officially, but it's also workable for users to override the pipeline as @ragulka showed. to get it working in the short term. 

Also, this kind of change should be non-breaking.

**Reply by Slund** [reply]
Any news on this? Could be awesome with support for precognition out of the box!

**Reply by ragulka** [reply]
One thing to note about the workaround above is that if you [overwrite the validator](https://spatie.be/docs/laravel-data/v4/validation/working-with-the-validator#content-overwriting-the-validator) in your data class to customize validation rules, then those modified rules won't be show up when calling `Data::getValidationRules`.

I discovered this by working on something different, but it seems like `Data::getValidationRules` and the actual data validation pipeline build validation rules using different mechanisms, which to me seems like a design issue, because there is no single source of truth.

**Reply by ragulka** [reply]
I found that overwriting `getValidationRules()` to use the `dataValidatorResolver` solves the issue with custom validators and ensures there's a single source of truth for validation rules:

```php
class MyData {

    public static function getValidationRules(array $payload): array
    {
        return DataContainer::get()->dataValidatorResolver()->execute(
            static::class,
            $payload,
        )->getRules();
    }
}
```

This is what the `ValidatableData::validate()` method does before actually validating the payload.

https://github.com/spatie/laravel-data/blob/48b3708f3911fa03306e718150d44286a1e8f422/src/Concerns/ValidateableData.php#L26-L29

@rubenvanassche I'm curious why `ValidatableData::getValidationRules` does not not use the same method? Some sort of overhead concern?

**Reply by rubenvanassche** [reply]
Yeah felt a bit useless to create a full validator and then only take the rules from it while we have some dedicated Resolvers for it which could be called separately.

That being said, I don't think this will be the biggest performance issue so if it can allow precognition to be added then you've got my go signal to rewrite it!

**Reply by valentinhttr** [reply]
Ok, I was thinking I had plenty of bugs with Laravel precognition before I realized that.... it is not support atm (my bad...).

So, I'm trying your solution @ragulka (`ValidatePropertiesDataPipe`).
But do you know if this is supposed to work with nested rules/data?

My precognitive header looks like:
``` 
property.0.start
```

But it has no effect on validation, 
```php
$payload->filterPrecognitiveRules(($class->name)::getValidationRules($properties)),
```
returns only first level rules (`property`).

Ay idea ?


EDIT:
overwriting `getValidationRules()` is the way to go for the moment.

**Reply by valentinhttr** [reply]
For those who might find it helpful, I had to include the Data's custom messages and attributes in the DataPipe, otherwise these are the default error messages.

```diff
public function handle(
        mixed $payload,
        DataClass $class,
        array $properties,
        CreationContext $creationContext
): array 
{
  if ($payload instanceof Request && $payload->isPrecognitive()) {
    Validator::make(
      $properties,
      $payload->filterPrecognitiveRules(($class->name)::getValidationRules($properties)),
+     ($class->name)::messages() ?? [],
+     ($class->name)::attributes() ?? [],
    )
      ->after(Precognition::afterValidationHook($payload))
      ->validate();
  }
  
  return parent::handle($payload, $class, $properties, $creationContext);
}

**Comment by patie** [comment]
It would be great if the support were built right into the package—any progress on that? Thanks
